### PR TITLE
Added integration testing util methods for WS unsubscribing and clean connection close

### PR DESCRIPTION
**Pull request by dimiter-todorov-georgiev**
[#185] Create integration test util methods for unsubscribing from WS messages, and for clean closing of WS connections

- UnsubscribeFromWSMessages
- Disconnect

Signed-off-by: Dimiter Georgiev <dimiter.georgiev@bosch.io>

### CODE REVIEW

File: integration/util/web.go [hunk 1]
```diff
@@ -137,11 +138,32 @@ func SubscribeForWSMessages(cfg *TestConfiguration, conn *websocket.Conn, eventT
...
+	if err := SubscribeForWSMessages(cfg, ws, messageType, ""); err != nil {
+		return fmt.Errorf("unable to unsubscribe with %s: %v", messageType, err)
+	}
+	return nil
+}
+
+// UnsubscribeFromWSMessages unsubscribes for the messages that are sent from a web socket session
+// and awaits confirmation response.
+// It logs any errors that occur, so that it can be called with defer.
+func UnsubscribeFromWSMessages(t *testing.T, cfg *TestConfiguration, ws *websocket.Conn, messageType string) {
```
Comment (gboyvalenkov-bosch) [inline integration/util/web.go:157]: Maybe we should return the error, not log it

File: integration/util/web.go [hunk 2]
```diff
@@ -137,11 +138,32 @@ func SubscribeForWSMessages(cfg *TestConfiguration, conn *websocket.Conn, eventT
 	return WaitForWSMessage(cfg, conn, fmt.Sprintf("%s:ACK", eventType))
 }
 
+func logError(t *testing.T, f func() error, description string) {
```
Comment (gboyvalenkov-bosch) [inline integration/util/web.go:141]: I think this is not needed

File: integration/util/web.go (hunk 1 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:157]: This is so we can call this with defer:

```
	ws, err := util.NewDigitalTwinWSConnection(suite.Cfg)
	require.NoError(suite.T(), err, "cannot create a websocket connection to the backend")
	defer Disconnect(suite.T(), suite.Cfg, ws)

	filter := fmt.Sprintf("like(resource:path,'/features/%s/*')", featureID)
	err = util.SubscribeForWSMessages(suite.Cfg, ws, "START-SEND-EVENTS", filter)
	require.NoError(suite.T(), err, "subscription for events should succeed")
	defer Unsubscribe(suite.T(), suite.Cfg, ws, "STOP-SEND-EVENTS")
```

File: integration/util/web.go (hunk 2 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:141]: See my reply on the `UnsubscribeFromWSMessages` comment.

Comment (gboyvalenkov-bosch) [reply integration/util/web.go:141]: Okay

File: integration/util/web.go (hunk 1 above)
Comment (hristobojilov) [inline integration/util/web.go:157]: Return errors if needed, do not log them like that.

**Review by hristobojilov — CHANGES_REQUESTED** [review]

File: integration/util/web.go (hunk 1 above)
Comment (hristobojilov) [reply integration/util/web.go:157]: Callers may choose to ignore errors returned and defer will work even if a method returns an error.

File: integration/util/web.go [hunk 3]
```diff
@@ -137,11 +137,22 @@ func SubscribeForWSMessages(cfg *TestConfiguration, conn *websocket.Conn, eventT
 	return WaitForWSMessage(cfg, conn, fmt.Sprintf("%s:ACK", eventType))
 }
 
+// UnsubscribeFromWSMessages unsubscribes for the messages that are sent from a web socket session
+// and awaits confirmation response.
+// It logs any errors that occur, so that it can be called with defer.
+func UnsubscribeFromWSMessages(cfg *TestConfiguration, ws *websocket.Conn, messageType string) error {
+	if err := SubscribeForWSMessages(cfg, ws, messageType, ""); err != nil {
```
Comment (e-grigorov) [inline integration/util/web.go:144]: Unsubscribe is using `subscribe`. That's not correct at all, it seems that some more refactoring is required.

File: integration/util/web.go [hunk 4]
```diff
@@ -161,7 +172,9 @@ func WaitForWSMessage(cfg *TestConfiguration, ws *websocket.Conn, expectedMessag
 func ProcessWSMessages(cfg *TestConfiguration, ws *websocket.Conn, process func(*protocol.Envelope) (bool, error)) error {
 	timeout := MillisToDuration(cfg.WsEventTimeoutMs)
 	deadline := time.Now().Add(timeout)
-	ws.SetDeadline(deadline)
+	if err := ws.SetDeadline(deadline); err != nil {
+		return fmt.Errorf("unable to set deadline to websocket: %v", err)
```
Comment (e-grigorov) [inline integration/util/web.go:176]: I just realized that WebSocket is written in a bunch of different ways. Could you find/replace all of them to: WebSocket?

File: integration/util/web.go [hunk 5]
```diff
@@ -190,6 +202,29 @@ func ProcessWSMessages(cfg *TestConfiguration, ws *websocket.Conn, process func(
 	return err
 }
 
+// Disconnect calls Close() on the websocket connection,
+// and then waits to ensure that the connection is closed.
+// It logs any errors that occur, so that it can be called with defer.
+func Disconnect(cfg *TestConfiguration, ws *websocket.Conn) error {
+	deadline := time.Now().Add(MillisToDuration(cfg.WsEventTimeoutMs))
```
Comment (e-grigorov) [inline integration/util/web.go:209]: Hm, it seems that we somehow missed that. The configuration property should be: WSEventTimeoutMS. MQTT names have to be fixed also.

File: integration/util/web.go [hunk 6]
```diff
@@ -190,6 +202,29 @@ func ProcessWSMessages(cfg *TestConfiguration, ws *websocket.Conn, process func(
...
+	}
+
+	if err := ws.Close(); err != nil {
+		return fmt.Errorf("unable to close websocket: %v", err)
+	}
+
+	var payload []byte
+	for time.Now().Before(deadline) {
+		if err := websocket.Message.Receive(ws, &payload); err != nil {
+			// We want this error, it shows that we've closed the connection
```
Comment (e-grigorov) [inline integration/util/web.go:221]: I am afraid that I cannot understand this comment. Could you please clarify the overall situation and how can this error help?

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: integration/util/web.go [hunk 7]
```diff
@@ -137,11 +137,22 @@ func SubscribeForWSMessages(cfg *TestConfiguration, conn *websocket.Conn, eventT
 	return WaitForWSMessage(cfg, conn, fmt.Sprintf("%s:ACK", eventType))
 }
 
+// UnsubscribeFromWSMessages unsubscribes for the messages that are sent from a web socket session
+// and awaits confirmation response.
+// It logs any errors that occur, so that it can be called with defer.
+func UnsubscribeFromWSMessages(cfg *TestConfiguration, ws *websocket.Conn, messageType string) error {
```
[the remark this answers is not in the document]
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:143]: It has to be `from`

File: integration/util/web.go (hunk 4 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:176]: I propose this naming convention:
- In comments: `WebSocket`
- In error messages: `websocket`
- Go Package name: `websocket`
- In variable names `ws`

File: integration/util/web.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go]: Done.

File: integration/util/web.go [hunk 8]
```diff
@@ -137,11 +137,22 @@ func SubscribeForWSMessages(cfg *TestConfiguration, conn *websocket.Conn, eventT
 	return WaitForWSMessage(cfg, conn, fmt.Sprintf("%s:ACK", eventType))
 }
 
+// UnsubscribeFromWSMessages unsubscribes for the messages that are sent from a web socket session
+// and awaits confirmation response.
+// It logs any errors that occur, so that it can be called with defer.
```
[the remark this answers is not in the document]
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go]: Done.

File: integration/util/web.go (hunk 3 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go]: Done.

File: integration/util/web.go (hunk 4 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:176]: Done.

File: integration/util/web.go (hunk 6 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go]: Done.

File: integration/util/web.go (hunk 5 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:208]: I believe we missed this because `golint` and `go vet` don't complain about these.

Comment (dimiter-todorov-georgiev) [reply integration/util/web.go]: Done.

**Review by hristobojilov — APPROVED** [review]

File: integration/util/web.go [hunk 9]
```diff
@@ -184,12 +196,41 @@ func ProcessWSMessages(cfg *TestConfiguration, ws *websocket.Conn, process func(
...
 	if !finished {
-		return fmt.Errorf("not finished, expected WS response not received in %v, last error: %v", timeout, err)
+		return fmt.Errorf("not finished, expected websocket response not received in %v, last error: %v", timeout, err)
 	}
 
 	return err
 }
 
+// Disconnect calls Close() on the WebSocket connection.
+// Then we wait until the connection is fully closed or the timeout expires.
```
Comment (e-grigorov) [inline integration/util/web.go:206]: `we wait` - API doc doesn't use such style. The text has to describe what the method does and there should not be: I, we etc.

File: integration/util/web.go [hunk 10]
```diff
@@ -184,12 +196,41 @@ func ProcessWSMessages(cfg *TestConfiguration, ws *websocket.Conn, process func(
...
+func Disconnect(cfg *TestConfiguration, ws *websocket.Conn) error {
+	deadline := time.Now().Add(MillisToDuration(cfg.WSEventTimeoutMS))
+	if err := ws.SetDeadline(deadline); err != nil {
+		return fmt.Errorf("unable to set deadline to websocket: %v", err)
+	}
+
+	// We call Close(), which sends a WebSocket close message to the backend.
```
Comment (e-grigorov) [inline integration/util/web.go:213]: `we call` - API doc doesn't use such style. The text has to describe what the method does and there should not be: I, we etc.

File: integration/util/web.go [hunk 11]
```diff
@@ -184,12 +196,41 @@ func ProcessWSMessages(cfg *TestConfiguration, ws *websocket.Conn, process func(
...
+	if err := ws.Close(); err != nil {
+		return fmt.Errorf("unable to close websocket: %v", err)
+	}
+
+	// Then the backend sends back a confirmation for the close message.
+	// We make sure we recieve any remaining messages from the server, including the confirmation,
+	// until we get an error, which means that the connection has been closed.
+	var payload []byte
+	for time.Now().Before(deadline) {
+		if err := websocket.Message.Receive(ws, &payload); err != nil {
```
Comment (e-grigorov) [inline integration/util/web.go:223]: If a concrete message is expected, it has to be asserted. Such generic error handling can be an issue.

**Review by hristobojilov — APPROVED** [review]

**Review by e-grigorov — APPROVED** [review]

**Pull request merged**